File: leetcode-solutions/python/133. Clone Graph.py
# https://leetcode.com/problems/clone-graph/

# by calling DFS two times
# first create all the nodes
# then push the neighbors
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        node_container = {}
        stack = [node]
        visited = {}
        visited[node.val] = True
        # create all the nodes and contain on the node_container
        while stack:
            cur_node = stack.pop()
            node_container[cur_node.val] = Node(cur_node.val)
            for neighbor in cur_node.neighbors:
                if neighbor.val in visited:
                    continue
                visited[neighbor.val] = True
                stack.append(neighbor)
        
        # assign all the neighbors
        stack = [node]
        visited = {}
        visited[node.val] = True
        while stack:
            cur_node = stack.pop()
            visited[cur_node.val] = True
            
            for neighbor in cur_node.neighbors:
                node_container[cur_node.val].neighbors.append(node_container[neighbor.val])
                if neighbor.val in visited:
                    continue
                stack.append(neighbor)
                visited[neighbor.val] = True
        return node_container[node.val]

File: leetcode-solutions/ruby/133. Clone Graph.rb
# https://leetcode.com/problems/clone-graph/

# Definition for a Node.
# class Node
#     attr_accessor :val, :neighbors
#     def initialize(val = 0, neighbors = nil)
#		  @val = val
#		  neighbors = [] if neighbors.nil?
#         @neighbors = neighbors
#     end
# end

# @param {Node} node
# @return {Node}
def cloneGraph(node)
    return node if node.nil?
    
    visited = {}
    stack = [node]
    container = {}
    visited[node.val] = true

    until stack.empty?
        cur = stack.pop
        unless container.key?(cur.val)
            container[cur.val] = Node.new(cur.val)
        end
        
        new_node = container[cur.val]
        
        cur.neighbors.each do |node|
            unless container.key?(node.val)
                container[node.val] = Node.new(node.val)
            end
            new_neighbor = container[node.val]
            new_neighbor.neighbors << new_node
            
            next if visited.key?(node.val)
            visited[node.val] = true
            stack << node
        end
    end

    container[node.val]
end
